### ISSUE TITLE
Consider using a memfd in the shmem library

### ISSUE THREAD

**Issue by stevenengler** [shadow/shadow#2901]
The shmem library uses `shm_open()` to create shared memory files in `/dev/shm`. These files are mapped into shadow and the shim.

We should consider using memfd files instead of `/dev/shm` filesystem-backed files now that all supported platforms support `memfd_create()`. This would allow us to remove Shadow's `shm_cleanup` code, and might also allow us to remove the Docker [requirements](https://shadow.github.io/docs/guide/supported_platforms.html#docker) for increasing the size of the `/dev/shm` mount (these errors have been confusing and difficult to debug in the past).

This would require us to change how the files are shared between shadow and the shim. Currently the file paths (`/dev/shm/something`) are sent to the shim as part of a serialized handle, and the shim opens the files and maps them into its process space. We would probably want to have the shim open them from `/proc/<shadow-pid>/fd/<fd>`, or better to pass them to the shim through the fork-exec process (disable `CLOEXEC` after the fork).

This would require us to simplify the shmem library. A current feature of the shmem library is that you could pass the serialized handle to any process, and that process could access the shared memory, but if we used a memfd we would/might lose that ability.

**Comment by sporksmith** [comment]
> might also allow us to remove the Docker [requirements](https://shadow.github.io/docs/guide/supported_platforms.html#docker) for increasing the size of the /dev/shm mount (these errors have been confusing and difficult to debug in the past).

I *think* the files are still created in `/dev/shm`; even if they're unlinked immediately, their storage comes from there.

> A current feature of the shmem library is that you could pass the serialized handle to any process, and that process could access the shared memory, but if we used a memfd we would/might lose that ability.

I think we could replace the current string path with `/proc/<shadow-pid>/fd/<fd>`; or more compactly just `<shadow-pid>,<fd>`

I have a WIP PR to make this change, but am currently debugging. Hopefully there's not a fundamental issue here that I'm missing...

**Comment by stevenengler** [comment]
> > might also allow us to remove the Docker [requirements](https://shadow.github.io/docs/guide/supported_platforms.html#docker) for increasing the size of the /dev/shm mount (these errors have been confusing and difficult to debug in the past).
> 
> I _think_ the files are still created in `/dev/shm`; even if they're unlinked immediately, their storage comes from there.

I don't think the memfd files are, for example `memfd_create()` works even if you don't have a `/dev/shm` mount.

**Comment by stevenengler** [comment]
> > A current feature of the shmem library is that you could pass the serialized handle to any process, and that process could access the shared memory, but if we used a memfd we would/might lose that ability.
> 
> I think we could replace the current string path with `/proc/<shadow-pid>/fd/<fd>`; or more compactly just `<shadow-pid>,<fd>`
> 
> I have a WIP PR to make this change, but am currently debugging. Hopefully there's not a fundamental issue here that I'm missing...

I think it would work, it's just sort of brittle. For example if shadow closes the fd and then the OS assigns a new file the same fd, the other process would map the wrong file. In the current shmem code the `open()` in the other process would at least fail in this case because of the random filenames. By passing file descriptors around explicitly, you can't run into this problem.

**Comment by sporksmith** [comment]
Good point. I suppose we could add a header in the file itself with a unique ID that's also included in the serialized handle.

**Comment by sporksmith** [comment]
Unassigned myself since @RWails is working on a Rust rewrite of this code
